### PR TITLE
Abifix

### LINKED ISSUE #608: bug in abiencoding
Currently, `seth calldata "f(string[],string)" '["hi there"]' '"asdlkhsdglakhgalksh"'` yields:
```
0x07ac501f000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000008686920746865726500000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000136173646c6b687364676c616b6867616c6b736800000000000000000000000000
```

while the correct result is:
```
0x07ac501f000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000c0000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000008686920746865726500000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000136173646c6b687364676c616b6867616c6b736800000000000000000000000000
```

### PR DESCRIPTION
**Pull request by MrChico**
based on top of #606. Fixes #608

### CODE REVIEW

File: src/hevm/CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,15 @@
...
   these should allow for much more ergonomic testing of code that handles signed messages.
 
+### Fixed
+
+- Fixed a bug in the abiencoding.
+- Fixed the range being generated by ints.
+
+### Changed
+
+- Arbitrary instance of AbiType can no longer a tuple
```
Comment (xwvvvvwx) [inline src/hevm/CHANGELOG.md:17]:
```suggestion
- Arbitrary instance of AbiType can no longer generate a tuple
```

File: src/hevm/test/test.hs [hunk 2]
```diff
@@ -638,6 +650,7 @@ hex s =
 singleContract :: Text -> Text -> IO (Maybe ByteString)
 singleContract x s =
   solidity x [i|
+    pragma experimental ABIEncoderV2;
```
Comment (xwvvvvwx) [inline src/hevm/test/test.hs:653]: maybe it's worth testing against abiencoderv1 as well?

File: src/hevm/src/EVM/ABI.hs [hunk 3]
```diff
@@ -455,22 +426,19 @@ genAbiValue = \case
     genUInt n = AbiUInt n <$> arbitraryIntegralWithMax (2^n-1)
 
 instance Arbitrary AbiType where
-  arbitrary = sized $ \n -> oneof $ -- prevent empty tuples
+  arbitrary = oneof $ -- doesn't create any tuples
```
Comment (xwvvvvwx) [inline src/hevm/src/EVM/ABI.hs:429]: why can't we make tuples?

File: src/hevm/src/EVM/ABI.hs [hunk 4]
```diff
@@ -278,22 +275,6 @@ putAbiTail x =
...
-    AbiBool _    -> 32
-    AbiArray _ _ xs -> Vector.sum (Vector.map abiHeadSize xs) +
-                       Vector.sum (Vector.map abiTailSize xs)
-    AbiBytesDynamic xs -> 32 + roundTo32Bytes (BS.length xs)
-    AbiArrayDynamic _ xs -> 32 + Vector.sum (Vector.map abiHeadSize xs) +
-                                Vector.sum (Vector.map abiTailSize xs)
-    AbiString s -> 32 + roundTo32Bytes (BS.length s)
-    AbiTuple v  -> sum (abiValueSize <$> v)
-
 abiTailSize :: AbiValue -> Int
```
Comment (xwvvvvwx) [inline src/hevm/src/EVM/ABI.hs:278]: what is the distinction between `abiHeadSize` and `abiTailSize`?

File: src/hevm/test/test.hs (hunk 2 above)
Comment (MrChico) [reply src/hevm/test/test.hs:661]: we can't test the interesting types like arrays with abiencoderv1

**Pull request merged**